src/main/scala/nodes/learning/KernelRidgeRegression.scala

```diff
@@ -0,0 +1,240 @@
+          // Subtract out K_bb * W_bb from residual
+          val rhs = y_bb - (residuals(l) - kernelBlockBlockMat.t * wBlockOld)
+
+          val wBlockNew = lhs \ rhs
+          wLocals(l)(block) = wBlockNew
+          val wBlockOldBC = labels.context.broadcast(wBlockOld)
+          val wBlockNewBC = labels.context.broadcast(wBlockNew)
+          (wBlockOldBC, wBlockNewBC)
+        }
+
+        var newModel = updateModel(
+          model, wBlockBCs.map(_._2), blockIdxsBC, preFixLengthsBC).cache()
+        // This is to truncate the lineage every 50 blocks
+        if (labels.context.getCheckpointDir.isDefined && block % 50 == 49) {
+          newModel = MatrixUtils.truncateLineage(newModel, false)
+        }
+
+        // materialize the new model
+        newModel.count()
+
+        model.unpersist(true)
+        model = newModel
+
+        logInfo(s"EPOCH_${pass}_BLOCK_${block} took " + 
+          ((System.nanoTime - blockBegin)/1e9) + " seconds")
+
+        // Unpersist the kernel block as we are done with it
+        kernelBlockMat.unpersist(true)
+
+        wBlockBCs.map { case (wBlockOldBC, wBlockNewBC) =>
+          wBlockOldBC.unpersist(true)
+          wBlockNewBC.unpersist(true)
+        }
+        blockIdxsBC.unpersist(true)
+      }
+    }
+    labelsMat.unpersist(true)
+    wLocals(0)
+  }
+
+  def updateModel(
+      model: RDD[Array[DenseMatrix[Double]]],
+      wBlockNewBC: Seq[Broadcast[DenseMatrix[Double]]],
+      blockIdxsBC: Broadcast[Array[Int]],
+      preFixLengthsBC: Broadcast[Array[Int]]): RDD[Array[DenseMatrix[Double]]] = {
+    val newModel = model.mapPartitionsWithIndex { case (idx, part) =>
+      // note that prefix length is *not* cumsum (so the first entry is 0)
+      val partBegin = preFixLengthsBC.value(idx) 
+      val wParts = part.next
+      assert(part.isEmpty)
+      wParts.zipWithIndex.foreach { case (wPart, idx) =>
+        val partLength = wPart.rows
+
+        // [partBegin, partBegin + partLength)
+        // [blockIdxs[0], blockIdxs[-1]]
+        //
+        // To compute which indices of the model to update, we take two
+        // ranges, map them both to example space ([nTrain]) so that we can
+        // intersect them, and then map the intersection back the delta model
+        // index space and partition index space.
+        val responsibleRange = (partBegin until (partBegin + partLength)).toSet
+        val inds = blockIdxsBC.value.zipWithIndex.filter { case (choice, _) => 
+          responsibleRange.contains(choice) 
+        }
+        val partInds = inds.map(x => x._1 - partBegin).toSeq
+        val blockInds = inds.map(x => x._2).toSeq
+
+        val wBlockNewBCvalue = wBlockNewBC(idx).value
+        wPart(partInds, ::) := wBlockNewBCvalue(blockInds, ::)
+      }
+      Iterator.single(wParts)
+    }
+    newModel
+  }
+
+  def addMatrixArrays(a: Array[DenseMatrix[Double]], b: Array[DenseMatrix[Double]]) = {
```

Comment:
Should this go in `MatrixUtils` (and be templated as `Seq[DenseMatrix[Double]]`?)